pkg/controller/apiextensions/crds/crd.go

```diff
@@ -0,0 +1,125 @@
+	"github.com/pkg/errors"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	kyaml "k8s.io/apimachinery/pkg/util/yaml"
+
+	"github.com/crossplane/crossplane/apis/apiextensions/v1alpha1"
+)
+
+// CRDOption is used to manipulate base crd.
+type CRDOption func(*v1beta1.CustomResourceDefinition)
+
+// NOTE(muvaf): We use v1beta1.CustomResourceDefinition for backward compatibility
+// with clusters pre-1.16
+
+// BaseCRD returns a base template for generating a CRD.
+func BaseCRD(opts ...CRDOption) *v1beta1.CustomResourceDefinition {
+	falseVal := false
+	// TODO(muvaf): Add proper descriptions.
+	crd := &v1beta1.CustomResourceDefinition{
+		Spec: v1beta1.CustomResourceDefinitionSpec{
+			PreserveUnknownFields: &falseVal,
+			Subresources: &v1beta1.CustomResourceSubresources{
+				Status: &v1beta1.CustomResourceSubresourceStatus{},
+			},
+			Validation: &v1beta1.CustomResourceValidation{
+				OpenAPIV3Schema: &v1beta1.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]v1beta1.JSONSchemaProps{
+						"apiVersion": {
+							Type: "string",
+						},
+						"kind": {
+							Type: "string",
+						},
+						"metadata": {
+							// NOTE(muvaf): api-server takes care of validating
+							// metadata.
+							Type: "object",
+						},
+						"spec": {
+							Type:       "object",
+							Properties: map[string]v1beta1.JSONSchemaProps{},
+						},
+						"status": {
+							Type:       "object",
+							Properties: map[string]v1beta1.JSONSchemaProps{},
+						},
+					},
+				},
+			},
+		},
+	}
+	for _, f := range opts {
+		f(crd)
+	}
+	return crd
+}
+
+// TODO(muvaf): move to a more programmatic approach where we have a func
+// that accepts a type and returns validation.
+
+// InfraValidation returns a CRDOption that adds infrastructure related fields
+// to the base CRD.
+func InfraValidation() CRDOption {
+	return func(crd *v1beta1.CustomResourceDefinition) {
+		crd.Spec.Scope = v1beta1.ClusterScoped
+		spec := &map[string]v1beta1.JSONSchemaProps{}
+		if err := yaml.Unmarshal([]byte(v1alpha1.InfraSpecProps), spec); err != nil {
+			// todo(muvaf): never panic.
+			panic(fmt.Sprintf("constant string could not be parsed: %s", err.Error()))
+		}
+		for k, v := range *spec {
+			crd.Spec.Validation.OpenAPIV3Schema.Properties["spec"].Properties[k] = v
+		}
+		status := &map[string]v1beta1.JSONSchemaProps{}
+		if err := yaml.Unmarshal([]byte(v1alpha1.InfraStatusProps), status); err != nil {
+			// todo(muvaf): never panic.
+			panic(fmt.Sprintf("constant string could not be parsed: %s", err.Error()))
+		}
+		for k, v := range *status {
+			crd.Spec.Validation.OpenAPIV3Schema.Properties["status"].Properties[k] = v
+		}
+	}
+}
+
+// GenerateInfraCRD returns a CRD that is generated with the information in cr.
+func GenerateInfraCRD(cr *v1alpha1.InfrastructureDefinition) (*v1beta1.CustomResourceDefinition, error) {
+	dec := kyaml.NewYAMLOrJSONDecoder(bytes.NewReader(cr.Spec.CRDSpecTemplate.Raw), 4096)
```

Comment:
Nit: might want to make 4096 a constant to make it obvious what it is (I'm guessing a buffer size?)

Reply:
> // NewYAMLOrJSONDecoder returns a decoder that will process YAML documents
// or JSON documents from the given reader as a stream. bufferSize determines
// how far into the stream the decoder will look to figure out whether this
// is a JSON stream (has whitespace followed by an open brace).

It is buffer size but it's only used to figure out whether it's JSON or YAML, sounded pretty unimportant in that case to make a variable on its own.